File: backend/app/api/routes/purchase_invoices.py
# Purchase invoices module: endpoints for viewing/editing/voiding a purchase
# invoice, and for attaching (or replacing) the vendor's own PDF. Purely a
# billing record: unlike create_new_purchase_order, nothing here touches
# inventory (see services/inventory.py::record_purchase_received) — that
# stays exclusively on the purchase-order-received flow. Restricted to admins
# (bypassed entirely when settings.auth_enabled is False, matching
# require_admin in routes/admin.py).
#
# There is deliberately no branded PDF here. A purchase invoice is a record
# of what a VENDOR billed us, so their own document is the authoritative one
# and the only one worth serving: rendering our own version of someone else's
# invoice produced a second document stating the same charge in our
# letterhead, which is not a thing we issue and not a thing anyone should
# file. get_purchase_invoice_uploaded_pdf is therefore the only PDF endpoint,
# and services/purchase_invoice_pdf.py is gone rather than left unused. Sales
# invoices still render (routes/invoices.py) — those are documents we
# genuinely issue.
#
# Nothing here creates a purchase invoice, and there is deliberately no
# endpoint that does: every purchase invoice is raised as part of creating
# the purchase order it belongs to (see create_new_purchase_order in
# routes/orders.py and services/purchase_invoices.py), whether that order was
# keyed in by hand or read off an uploaded vendor invoice PDF. A purchase
# invoice with no order behind it was never a valid record, so the create
# call that allowed one is gone rather than guarded.
#
# The vendor's own PDF is attached afterwards as its own request
# (attach_purchase_invoice_pdf) — same two-phase pattern as
# catalogues/products (see routes/catalogues.py's module docstring for why a
# file never travels bundled into another request). That endpoint doubles as
# "replace": if the invoice already has an uploaded PDF, the old file is
# hard-deleted from disk once the new one is saved and the record updated, so
# no history/versions are kept.
import io
import zipfile
from datetime import date, datetime, time

from fastapi import APIRouter, Depends, File, Form, HTTPException, Response, UploadFile, status

from app.api.routes.admin import require_admin
from app.models import PurchaseInvoiceDetails, User
from app.schemas.purchase_invoices import (
    AttachPurchaseInvoicePdfResponse,
    PurchaseInvoiceDetailItem,
    UpdatePurchaseInvoiceDetailsRequest,
    UpdatePurchaseInvoiceDetailsResponse,
)
from app.services.invoice_numbering import format_purchase_invoice_no
from app.services.purchase_invoice_storage import delete_uploaded_pdf, read_uploaded_pdf, save_uploaded_pdf
from app.services.storage import LocalUploadBlockedError

router = APIRouter(prefix="/admin", tags=["purchase-invoices"])


async def _get_purchase_invoice_or_404(purchase_invoice_id: int) -> PurchaseInvoiceDetails:
    purchase_invoice = await PurchaseInvoiceDetails.get(purchase_invoice_id)
    if purchase_invoice is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="purchase invoice not found")
    return purchase_invoice


@router.post("/attach_purchase_invoice_pdf", response_model=AttachPurchaseInvoicePdfResponse)
async def attach_purchase_invoice_pdf(
    purchase_invoice_id: int = Form(...),
    file: UploadFile = File(...),
    _: User | None = Depends(require_admin),
) -> AttachPurchaseInvoicePdfResponse:
    purchase_invoice = await _get_purchase_invoice_or_404(purchase_invoice_id)

    # Captured before overwriting below — if this invoice already has an
    # uploaded PDF, this call replaces it: the old file is only removed once
    # the new one is safely saved and the record points at it, so a failure
    # partway through never loses the previous PDF.
    previous_pdf_path = purchase_invoice.uploaded_pdf_path

    pdf_bytes = await file.read()
    try:
        purchase_invoice.uploaded_pdf_path = save_uploaded_pdf(pdf_bytes)
    except LocalUploadBlockedError as error:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(error))
    await purchase_invoice.save()

    if previous_pdf_path is not None:
        delete_uploaded_pdf(previous_pdf_path)

    return AttachPurchaseInvoicePdfResponse(message="PDF attached successfully")


def _to_purchase_invoice_detail_item(purchase_invoice: PurchaseInvoiceDetails) -> PurchaseInvoiceDetailItem:
    return PurchaseInvoiceDetailItem(
        id=purchase_invoice.id,
        purchase_invoice_no=purchase_invoice.purchase_invoice_no,
        purchase_invoice_no_display=format_purchase_invoice_no(purchase_invoice.purchase_invoice_no),
        date=purchase_invoice.date,
        vendor_id=purchase_invoice.vendor_id,
        po_id=purchase_invoice.po_id,
        vendor_invoice_no=purchase_invoice.vendor_invoice_no,
        has_uploaded_pdf=purchase_invoice.uploaded_pdf_path is not None,
        total_amount_before_tax=purchase_invoice.total_amount_before_tax,
        total_tax_amount=purchase_invoice.total_tax_amount,
        total_amount_after_tax=purchase_invoice.total_amount_after_tax,
        is_deleted=purchase_invoice.is_deleted,
    )


@router.get("/get_purchase_invoice_details", response_model=list[PurchaseInvoiceDetailItem])
async def get_purchase_invoice_details(
    _: User | None = Depends(require_admin),
) -> list[PurchaseInvoiceDetailItem]:
    purchase_invoices = await PurchaseInvoiceDetails.find(PurchaseInvoiceDetails.is_deleted == False).to_list()
    return [_to_purchase_invoice_detail_item(purchase_invoice) for purchase_invoice in purchase_invoices]


@router.post("/update_purchase_invoice_details", response_model=UpdatePurchaseInvoiceDetailsResponse)
async def update_purchase_invoice_details(
    payload: UpdatePurchaseInvoiceDetailsRequest,
    _: User | None = Depends(require_admin),
) -> UpdatePurchaseInvoiceDetailsResponse:
    purchase_invoice = await _get_purchase_invoice_or_404(payload.id)

    purchase_invoice.date = payload.date
    purchase_invoice.is_deleted = payload.is_deleted
    await purchase_invoice.save()

    return UpdatePurchaseInvoiceDetailsResponse(message="purchase invoice updated successfully")


@router.get("/get_purchase_invoice_uploaded_pdf")
async def get_purchase_invoice_uploaded_pdf(
    purchase_invoice_id: int,
    _: User | None = Depends(require_admin),
) -> Response:
    purchase_invoice = await _get_purchase_invoice_or_404(purchase_invoice_id)
    if purchase_invoice.uploaded_pdf_path is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="no uploaded PDF for this purchase invoice")

    pdf_bytes = read_uploaded_pdf(purchase_invoice.uploaded_pdf_path)
    if pdf_bytes is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="uploaded PDF file is missing")

    # Named after our own record rather than the vendor's file, so a folder
    # of downloads sorts and reads the way the invoices list does. It used to
    # carry an "-original" suffix, back when there was a generated PDF for it
    # to be distinguished from.
    display_no = format_purchase_invoice_no(purchase_invoice.purchase_invoice_no)
    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={"Content-Disposition": f'attachment; filename="{display_no}.pdf"'},
    )


@router.get("/get_purchase_invoices_pdf_zip")
async def get_purchase_invoices_pdf_zip(
    start_date: date,
    end_date: date,
    _: User | None = Depends(require_admin),
) -> Response:
    # Bulk counterpart of get_purchase_invoice_uploaded_pdf, mirroring
    # get_invoices_pdf_zip in routes/invoices.py: same inclusive-both-ends
    # date window, same one-file-per-invoice zip. What differs is what goes
    # in it — there is no branded render on this side, so the archive can
    # only carry the vendor PDFs that were actually attached. Invoices with
    # nothing attached (or whose file has gone missing from disk) are simply
    # left out rather than failing the whole download, matching the list,
    # which shows a dash for them instead of a Download button.
    start_dt = datetime.combine(start_date, time.min)
    end_dt = datetime.combine(end_date, time.max)

    purchase_invoices = await PurchaseInvoiceDetails.find(
        PurchaseInvoiceDetails.is_deleted == False,  # noqa: E712 — Beanie needs the comparison
        PurchaseInvoiceDetails.date >= start_dt,
        PurchaseInvoiceDetails.date <= end_dt,
    ).to_list()

    buffer = io.BytesIO()
    written = 0
    with zipfile.ZipFile(buffer, "w", zipfile.ZIP_DEFLATED) as archive:
        for purchase_invoice in purchase_invoices:
            if purchase_invoice.uploaded_pdf_path is None:
                continue
            pdf_bytes = read_uploaded_pdf(purchase_invoice.uploaded_pdf_path)
            if pdf_bytes is None:
                continue
            display_no = format_purchase_invoice_no(purchase_invoice.purchase_invoice_no)
            archive.writestr(f"{display_no}.pdf", pdf_bytes)
            written += 1

    if written == 0:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="no purchase invoice PDFs found in that date range",
        )

    zip_filename = f"purchase-invoices-{start_date.isoformat()}-to-{end_date.isoformat()}.zip"
    return Response(
        content=buffer.getvalue(),
        media_type="application/zip",
        headers={"Content-Disposition": f'attachment; filename="{zip_filename}"'},
    )

File: frontend/src/lib/purchase-invoices.ts
// ---------------------------------------------------------------------------
// Purchase invoice data for the /admin/invoices Purchase Invoices tab
// ---------------------------------------------------------------------------
// Mirrors lib/invoices.ts. Fetches from GET /admin/get_purchase_invoice_details
// (backend/app/api/routes/purchase_invoices.py), which only ever returns
// active (non-deleted) purchase invoices. A purchase invoice is always
// raised against an existing PurchaseOrders record — vendor and totals are
// snapshotted from it at create time, same borrowing convention as sales
// invoices riding on a sales order. poId/vendorId are immutable once raised;
// a vendor PDF can optionally be attached (and later replaced) independently
// of that via attachPurchaseInvoicePdf.
//
// There is deliberately no create function here: a purchase invoice is
// raised automatically as part of creating its purchase order (see
// createPurchaseOrder's flow in components/admin/purchase-order-form-modal.tsx
// and services/purchase_invoices.py on the backend), so the Purchase
// Invoices tab lists, edits and voids them but never adds one.
//
// Nor is there a branded PDF to download. A purchase invoice records what a
// VENDOR billed us, so their own document is the authoritative one and the
// only one served — downloadUploadedPurchaseInvoicePdf is the whole of it.
// Sales invoices still render (lib/invoices.ts); those are documents we
// genuinely issue.
import { apiFetch } from "@/lib/api";

export type PurchaseInvoice = {
  id: number;
  purchaseInvoiceNo: number;
  purchaseInvoiceNoDisplay: string;
  date: string;
  vendorId: number;
  poId: number;
  // The vendor's own invoice number, present only for orders created by
  // uploading their invoice PDF.
  vendorInvoiceNo: string | null;
  hasUploadedPdf: boolean;
  totalAmountBeforeTax: number;
  totalTaxAmount: number;
  totalAmountAfterTax: number;
  isDeleted: boolean;
};

// Shape returned by the backend's PurchaseInvoiceDetailItem schema.
type PurchaseInvoiceDetailItem = {
  id: number;
  purchase_invoice_no: number;
  purchase_invoice_no_display: string;
  date: string;
  vendor_id: number;
  po_id: number;
  vendor_invoice_no: string | null;
  has_uploaded_pdf: boolean;
  total_amount_before_tax: number;
  total_tax_amount: number;
  total_amount_after_tax: number;
  is_deleted: boolean;
};

function toPurchaseInvoice(item: PurchaseInvoiceDetailItem): PurchaseInvoice {
  return {
    id: item.id,
    purchaseInvoiceNo: item.purchase_invoice_no,
    purchaseInvoiceNoDisplay: item.purchase_invoice_no_display,
    date: item.date,
    vendorId: item.vendor_id,
    poId: item.po_id,
    vendorInvoiceNo: item.vendor_invoice_no,
    hasUploadedPdf: item.has_uploaded_pdf,
    totalAmountBeforeTax: item.total_amount_before_tax,
    totalTaxAmount: item.total_tax_amount,
    totalAmountAfterTax: item.total_amount_after_tax,
    isDeleted: item.is_deleted,
  };
}

export async function fetchPurchaseInvoices(): Promise<PurchaseInvoice[]> {
  const response = await apiFetch("/admin/get_purchase_invoice_details");
  if (!response.ok) {
    throw new Error("Failed to load purchase invoices");
  }

  const items: PurchaseInvoiceDetailItem[] = await response.json();
  return items.map(toPurchaseInvoice);
}

// Attaches a vendor PDF to a purchase invoice via POST
// /admin/attach_purchase_invoice_pdf, or replaces the one already attached —
// the backend hard-deletes the old file from disk once the new one is saved,
// so there's no history/versions kept.
export async function attachPurchaseInvoicePdf(purchaseInvoiceId: number, file: File): Promise<void> {
  const formData = new FormData();
  formData.append("purchase_invoice_id", String(purchaseInvoiceId));
  formData.append("file", file);

  const response = await apiFetch("/admin/attach_purchase_invoice_pdf", {
    method: "POST",
    body: formData,
  });

  if (!response.ok) {
    throw new Error("Failed to attach PDF");
  }
}

export type UpdatePurchaseInvoicePayload = {
  id: number;
  date: string;
  isDeleted: boolean;
};

export async function updatePurchaseInvoice(payload: UpdatePurchaseInvoicePayload): Promise<Response> {
  return apiFetch("/admin/update_purchase_invoice_details", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      id: payload.id,
      date: payload.date,
      is_deleted: payload.isDeleted,
    }),
  });
}

// The vendor's own PDF, and the only one there is. Named after our record
// rather than their file so a folder of downloads sorts the way the list
// does; it used to carry an "-original" suffix, back when there was a
// generated PDF for it to be distinguished from.
export async function downloadUploadedPurchaseInvoicePdf(
  purchaseInvoiceId: number,
  purchaseInvoiceNoDisplay: string,
): Promise<void> {
  const response = await apiFetch(
    `/admin/get_purchase_invoice_uploaded_pdf?purchase_invoice_id=${purchaseInvoiceId}`,
  );
  if (!response.ok) {
    throw new Error("Failed to download PDF");
  }

  const blob = await response.blob();
  const url = URL.createObjectURL(blob);
  const link = document.createElement("a");
  link.href = url;
  link.download = `${purchaseInvoiceNoDisplay}.pdf`;
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
  URL.revokeObjectURL(url);
}

// Bulk-downloads every attached vendor PDF dated within [startDate, endDate]
// (both "YYYY-MM-DD") as a single .zip — the purchase-side counterpart of
// downloadInvoicesZip in lib/invoices.ts, and the same blob-and-throwaway-link
// approach, since the endpoint needs the Authorization header too. Invoices
// with no PDF attached are skipped by the backend rather than erroring, so a
// range containing only those comes back as the 404 below.
export async function downloadPurchaseInvoicesZip(startDate: string, endDate: string): Promise<void> {
  const response = await apiFetch(
    `/admin/get_purchase_invoices_pdf_zip?start_date=${encodeURIComponent(startDate)}&end_date=${encodeURIComponent(endDate)}`,
  );
  if (!response.ok) {
    if (response.status === 404) {
      throw new Error("No purchase invoice PDFs found in that date range.");
    }
    throw new Error("Failed to generate purchase invoices zip");
  }

  const blob = await response.blob();
  const url = URL.createObjectURL(blob);
  const link = document.createElement("a");
  link.href = url;
  link.download = `purchase-invoices-${startDate}-to-${endDate}.zip`;
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
  URL.revokeObjectURL(url);
}
